lib/pages/RegistrationForm.dart

```diff
@@ -0,0 +1,215 @@
+import 'package:cryptography/cryptography.dart';
+import 'package:hive/hive.dart';
+
+import 'HomePages.dart';
+import 'SplashScreen.dart';
+
+Future<Map> _calculateEncryptionValues(String pin) async {
+  HashMap values = HashMap();
+
+  // this should give us 256 bits? someone check my math
+  final SecretKey randKey = SecretKey(
+      List<int>.generate(32, (index) => Random.secure().nextInt(255)));
+  final List<int> randKeyBytes = await randKey.extractBytes();
+  values.putIfAbsent("randKey", () => randKeyBytes);
+
+  // utf8 or ascii here? I guess utf8 gives us a bigger PIN space?
+  final SecretKey pinBytes = SecretKey(utf8.encode(pin));
+
+  final pbkdf2 = Pbkdf2(
+    macAlgorithm: Hmac.sha512(),
+    iterations: 150000,
+    bits: 256,
+  );
+
+  final SecretKey salt = SecretKey(
+      List<int>.generate(16, (index) => Random.secure().nextInt(255)));
+  final String saltStr = hex.encode(await salt.extractBytes());
+  values.putIfAbsent("salt", () => saltStr);
+
+  final userKey = await pbkdf2.deriveKey(
+    secretKey: pinBytes,
+    nonce: await salt.extractBytes(),
+  );
+
+  final sha512 = Sha512();
+  final hashedUserKey = await sha512.hash(await userKey.extractBytes());
+  final String hashedUserKeyStr = hex.encode(hashedUserKey.bytes);
+  values.putIfAbsent("hashedUserKey", () => hashedUserKeyStr);
+
+  final aes256 = AesGcm.with256bits();
+  final aesNonce = aes256.newNonce();
+
+  final encryptedRandKey = await aes256.encrypt(
+    await randKey.extractBytes(),
+    secretKey: userKey,
+    nonce: aesNonce,
+  );
+  values.putIfAbsent(
+      "randKeyCipherText", () => hex.encode(encryptedRandKey.cipherText));
+  values.putIfAbsent("randKeyNonce", () => hex.encode(encryptedRandKey.nonce));
+  values.putIfAbsent(
+      "randKeyMAC", () => hex.encode(encryptedRandKey.mac.bytes));
+
+  return values;
+}
+
+Future<Box> _registerPIN(String pin) async {
+  final encryptionValues = await compute(_calculateEncryptionValues, pin);
+  SharedPreferences prefs = await SharedPreferences.getInstance();
+
+  final List<int> randKeyBytes = encryptionValues["randKey"];
+  final calendarBox = await Hive.openBox('calendarBox', encryptionCipher: HiveAesCipher(randKeyBytes));
+
+  prefs.setString("salt", encryptionValues["salt"]);
+  prefs.setString("hashedUserKey", encryptionValues["hashedUserKey"]);
+
+  prefs.setString("randKeyCipherText", encryptionValues["randKeyCipherText"]);
+  prefs.setString("randKeyNonce", encryptionValues["randKeyNonce"]);
+  prefs.setString("randKeyMAC", encryptionValues["randKeyMAC"]);
+
+  prefs.setBool("onboarded", true);
+
+  return calendarBox;
+}
+
+class RegistrationLoadingPage extends StatefulWidget {
+  final String pin;
+
+  const RegistrationLoadingPage({Key? key, required this.pin})
+      : super(key: key);
+
+  @override
+  State<RegistrationLoadingPage> createState() =>
+      _RegistrationLoadingPageState();
+}
+
+class _RegistrationLoadingPageState extends State<RegistrationLoadingPage> {
```

Comment:
This would be another candidate for an abstracted implementation of the splash screen, if you can come up with a good way to do it.

Reply:
Resolved by [0f0fe3e](https://github.com/Yours-App/Secure-Period-Tracker/pull/10/commits/0f0fe3ee301d88cd6ee8103d5a7835c319cd6163)